### ISSUE TITLE
[jvm-packages] - sample code error with 'SparkContext has been shutdown' on AWS EMR

### ISSUE THREAD

**Issue by dss99911** [dmlc/xgboost#7794]
I ran the [example code](https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-example/src/main/scala/ml/dmlc/xgboost4j/scala/example/spark/SparkTraining.scala) on spark 3.1.2, xgboost4j-spark 1.5.2 on Mac local mode.
and it worked fine

but errored on AWS EMR 6.5.0, spark 3.1.2, xgboost4j-spark 1.5.2 with the error below
the error looks similar with https://github.com/dmlc/xgboost/issues/7700

but my error log shows `SyntaxError: invalid syntax` on TrackerProcessLogger. so, created new issue.
if it's same issue. I'll close this issue.

```
22/04/10 19:58:33 WARN XGBoostSpark: train_test_ratio is deprecated since XGBoost 0.82, we recommend to explicitly pass a training and multiple evaluation datasets by passing 'eval_sets' and 'eval_set_names'
22/04/10 19:58:33 INFO RabitTracker$TrackerProcessLogger:   File "/tmp/tracker3786293075004416443.py", line 67
22/04/10 19:58:33 INFO RabitTracker$TrackerProcessLogger:     assert magic == kMagic, f"invalid magic number={magic} from {self.host}"
22/04/10 19:58:33 INFO RabitTracker$TrackerProcessLogger:                                                                            ^
22/04/10 19:58:33 INFO RabitTracker$TrackerProcessLogger: SyntaxError: invalid syntax
22/04/10 19:58:33 INFO RabitTracker$TrackerProcessLogger: Tracker Process ends with exit code 1
22/04/10 19:58:33 INFO XGBoostSpark: starting training with timeout set as 600000 ms for waiting for resources
22/04/10 19:58:33 INFO RabitTracker: Tracker Process ends with exit code 1
22/04/10 19:58:33 INFO XGBoostSpark: Rabit returns with exit code 1
22/04/10 19:58:33 ERROR XGBoostSpark: the job was aborted due to 
ml.dmlc.xgboost4j.java.XGBoostError: XGBoostModel training failed
	at ml.dmlc.xgboost4j.scala.spark.XGBoost$.postTrackerReturnProcessing(XGBoost.scala:750)
	at ml.dmlc.xgboost4j.scala.spark.XGBoost$.trainDistributed(XGBoost.scala:624)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.train(XGBoostClassifier.scala:199)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.train(XGBoostClassifier.scala:40)
	at org.apache.spark.ml.Predictor.fit(Predictor.scala:151)
	at org.apache.spark.ml.Predictor.fit(Predictor.scala:115)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$5(Pipeline.scala:151)
	at org.apache.spark.ml.MLEvents.withFitEvent(events.scala:130)
	at org.apache.spark.ml.MLEvents.withFitEvent$(events.scala:123)
	at org.apache.spark.ml.util.Instrumentation.withFitEvent(Instrumentation.scala:42)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$4(Pipeline.scala:151)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$2(Pipeline.scala:147)
	at org.apache.spark.ml.MLEvents.withFitEvent(events.scala:130)
	at org.apache.spark.ml.MLEvents.withFitEvent$(events.scala:123)
	at org.apache.spark.ml.util.Instrumentation.withFitEvent(Instrumentation.scala:42)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$1(Pipeline.scala:133)
	at org.apache.spark.ml.util.Instrumentation$.$anonfun$instrumented$1(Instrumentation.scala:191)
	at scala.util.Try$.apply(Try.scala:213)
	at org.apache.spark.ml.util.Instrumentation$.instrumented(Instrumentation.scala:191)
	at org.apache.spark.ml.Pipeline.fit(Pipeline.scala:133)
	at com.balancehero.acs.TransactionPredictionModel2$.main(TransactionPredictionModel2.scala:107)
	at com.balancehero.acs.TransactionPredictionModel2.main(TransactionPredictionModel2.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:959)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1047)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1056)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
22/04/10 19:58:33 ERROR RabitTracker: Uncaught exception thrown by worker:
java.lang.IllegalStateException: SparkContext has been shutdown
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2233)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2262)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2281)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2306)
	at org.apache.spark.rdd.RDD.$anonfun$foreachPartition$1(RDD.scala:1020)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:414)
	at org.apache.spark.rdd.RDD.foreachPartition(RDD.scala:1018)
	at ml.dmlc.xgboost4j.scala.spark.XGBoost$$anon$1.run(XGBoost.scala:612)
22/04/10 19:58:33 ERROR RabitTracker: java.lang.InterruptedException: sleep interrupted
22/04/10 19:58:33 INFO AbstractConnector: Stopped Spark@18989e0{HTTP/1.1, (http/1.1)}{0.0.0.0:4040}
22/04/10 19:58:33 INFO SparkUI: Stopped Spark web UI at http://ip-10-50-6-154.ap-south-1.compute.internal:4040
22/04/10 19:58:33 INFO YarnClientSchedulerBackend: Interrupting monitor thread
22/04/10 19:58:33 INFO YarnClientSchedulerBackend: Shutting down all executors
22/04/10 19:58:33 INFO YarnSchedulerBackend$YarnDriverEndpoint: Asking each executor to shut down
22/04/10 19:58:33 INFO YarnClientSchedulerBackend: YARN client scheduler backend Stopped
22/04/10 19:58:33 INFO MapOutputTrackerMasterEndpoint: MapOutputTrackerMasterEndpoint stopped!
22/04/10 19:58:33 INFO MemoryStore: MemoryStore cleared
22/04/10 19:58:33 INFO BlockManager: BlockManager stopped
22/04/10 19:58:33 INFO BlockManagerMaster: BlockManagerMaster stopped
22/04/10 19:58:33 INFO OutputCommitCoordinator$OutputCommitCoordinatorEndpoint: OutputCommitCoordinator stopped!
22/04/10 19:58:33 INFO SparkContext: Successfully stopped SparkContext
22/04/10 19:58:33 ERROR Instrumentation: ml.dmlc.xgboost4j.java.XGBoostError: XGBoostModel training failed
	at ml.dmlc.xgboost4j.scala.spark.XGBoost$.postTrackerReturnProcessing(XGBoost.scala:750)
	at ml.dmlc.xgboost4j.scala.spark.XGBoost$.trainDistributed(XGBoost.scala:624)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.train(XGBoostClassifier.scala:199)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.train(XGBoostClassifier.scala:40)
	at org.apache.spark.ml.Predictor.fit(Predictor.scala:151)
	at org.apache.spark.ml.Predictor.fit(Predictor.scala:115)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$5(Pipeline.scala:151)
	at org.apache.spark.ml.MLEvents.withFitEvent(events.scala:130)
	at org.apache.spark.ml.MLEvents.withFitEvent$(events.scala:123)
	at org.apache.spark.ml.util.Instrumentation.withFitEvent(Instrumentation.scala:42)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$4(Pipeline.scala:151)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$2(Pipeline.scala:147)
	at org.apache.spark.ml.MLEvents.withFitEvent(events.scala:130)
	at org.apache.spark.ml.MLEvents.withFitEvent$(events.scala:123)
	at org.apache.spark.ml.util.Instrumentation.withFitEvent(Instrumentation.scala:42)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$1(Pipeline.scala:133)
	at org.apache.spark.ml.util.Instrumentation$.$anonfun$instrumented$1(Instrumentation.scala:191)
	at scala.util.Try$.apply(Try.scala:213)
	at org.apache.spark.ml.util.Instrumentation$.instrumented(Instrumentation.scala:191)
	at org.apache.spark.ml.Pipeline.fit(Pipeline.scala:133)
	at com.balancehero.acs.TransactionPredictionModel2$.main(TransactionPredictionModel2.scala:107)
	at com.balancehero.acs.TransactionPredictionModel2.main(TransactionPredictionModel2.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:959)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1047)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1056)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)

Exception in thread "main" ml.dmlc.xgboost4j.java.XGBoostError: XGBoostModel training failed
	at ml.dmlc.xgboost4j.scala.spark.XGBoost$.postTrackerReturnProcessing(XGBoost.scala:750)
	at ml.dmlc.xgboost4j.scala.spark.XGBoost$.trainDistributed(XGBoost.scala:624)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.train(XGBoostClassifier.scala:199)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier.train(XGBoostClassifier.scala:40)
	at org.apache.spark.ml.Predictor.fit(Predictor.scala:151)
	at org.apache.spark.ml.Predictor.fit(Predictor.scala:115)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$5(Pipeline.scala:151)
	at org.apache.spark.ml.MLEvents.withFitEvent(events.scala:130)
	at org.apache.spark.ml.MLEvents.withFitEvent$(events.scala:123)
	at org.apache.spark.ml.util.Instrumentation.withFitEvent(Instrumentation.scala:42)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$4(Pipeline.scala:151)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$2(Pipeline.scala:147)
	at org.apache.spark.ml.MLEvents.withFitEvent(events.scala:130)
	at org.apache.spark.ml.MLEvents.withFitEvent$(events.scala:123)
	at org.apache.spark.ml.util.Instrumentation.withFitEvent(Instrumentation.scala:42)
	at org.apache.spark.ml.Pipeline.$anonfun$fit$1(Pipeline.scala:133)
	at org.apache.spark.ml.util.Instrumentation$.$anonfun$instrumented$1(Instrumentation.scala:191)
	at scala.util.Try$.apply(Try.scala:213)
	at org.apache.spark.ml.util.Instrumentation$.instrumented(Instrumentation.scala:191)
	at org.apache.spark.ml.Pipeline.fit(Pipeline.scala:133)
	at com.balancehero.acs.TransactionPredictionModel2$.main(TransactionPredictionModel2.scala:107)
	at com.balancehero.acs.TransactionPredictionModel2.main(TransactionPredictionModel2.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:959)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1047)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1056)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
22/04/10 19:58:33 INFO ShutdownHookManager: Shutdown hook called
22/04/10 19:58:33 INFO ShutdownHookManager: Deleting directory /mnt/tmp/spark-b3fd370a-f1ff-4ebd-b110-9b468d0cd616
22/04/10 19:58:33 INFO ShutdownHookManager: Deleting directory /mnt/tmp/spark-73a8e00b-3c1a-437a-aa3b-45f3b38bb8bc
Command exiting with ret '1'

```

**Comment by trivialfis** [comment]
Could you please specify the Python version on your driver node?  XGBoost requires Python >= 3.7 .

**Comment by dss99911** [comment]
the python version is 3.7.9

**Comment by trivialfis** [comment]
@dss99911 I think a double-check might help:

```
22/04/10 19:58:33 INFO RabitTracker$TrackerProcessLogger:   File "/tmp/tracker3786293075004416443.py", line 67
22/04/10 19:58:33 INFO RabitTracker$TrackerProcessLogger:     assert magic == kMagic, f"invalid magic number={magic} from {self.host}"
```

That's an error about the use of `f-string` in Python, which is available after Python 3.6.

**Comment by dss99911** [comment]
@trivialfis I double-checked and found the below. please review


driver node shell
```
[hadoop@ip-xx-xx-xx-xx ~]$ python -V
Python 3.7.9
[hadoop@ip-xx-xx-xx-xx ~]$ python3 -V
Python 3.7.9
[hadoop@ip-xx-xx-xx-xx~]$ which python
alias python='/usr/bin/python3'
	/usr/bin/python3
```

spark-shell
```
scala> "python -V".!!
Python 2.7.18
res1: String = ""

scala> "python3 -V".!!
res2: String =
"Python 3.7.9
"

scala> "which python".!
/usr/bin/python
```

pyspark also uses Python 3.7.9

I think the AWS EMR contains python2 and python3 both. but the python command is aliased to python3. so, pyspark is running on python3
but, on scala, python is not aliased. so, it uses python2.

I'm not sure what is the best approach.
but, [spark environment variable](https://spark.apache.org/docs/latest/configuration.html#environment-variables) doesn't support Scala python path. so, it seems not easy to configure this library to call python3.
Please share your opinion.

**Comment by dss99911** [comment]
when I create EMR cluster by this tutorial(https://nvidia.github.io/spark-rapids/docs/get-started/getting-started-aws-emr.html)
it's working fine.
I think something related to jar(I used fatjar previously) or some configuration.
as it's solved by this way. I'll close the issue.
thanks for the support

**Comment by wbo4958** [comment]
@dss99911 

 I just had a PR to fix this issue by specifying the rabit tracker host IP

``` scala
val hostIP = xx.xx.xx.xx
"tracker_conf" -> TrackerConf(0L, "python", hostIp)
```

The PR will be shipped in 1.6.0 release.
